### ISSUE TITLE
Redirected with incorrect `client_id` resulting in GitHub 404 error page

### ISSUE THREAD

**Issue by andrewphilipsmith** [gocd-contrib/github-oauth-authorization-plugin#82]
I am attempting to configure this plugin in a new GoCD server. I've not been able to install/configure this plugin correctly.

Actual behaviour
---
When I go to my GoCD home page and press the "Login using GitHub OAuth authorization plugin" button I am redirected to this URL:

https://github.com/login/oauth/authorize?client_id={long-string-that-isn't-the-client-id}&redirect_uri={correct-url-encoded-callback-url}scope=user%3Aemail

This displays GitHub's 404 error page.

The only information that appear in the logs, is in `plugin-cd.go.authorization.github.log`:
```
2019-10-10 12:27:00,325 DEBUG [qtp2061494383-41] GitHubPlugin:46 - [Get Authorization Server URL] Getting authorization server url from auth config.
```

The `long-string-that-isn't-the-client-id` is url-encoded, has the prefix `AES:` and *looks like* it could be an encrypted representation of the client_id. I don't know if or how I could confirm that.

If I manually edit the url in the browser and replace the `long-string-that-isn't-the-client-id` with the actual, plain-text `client_id` I am redirected to Github's Authorization page "App-X want to access to access your ... account", which is the desired result.

Expected behaviour
---
When I go to my GoCD home page and press the "Login using GitHub OAuth authorization plugin" button I expect to be redirected to Github's Authorization page "App-X want to access to access your ... account".

How to reproduce
---
Details from https://{server-url}/go/about:
```
Go Server Version: 19.9.0 (10194-34fca8f7682a65ef5fbd6726d51fd52b0c041bc6)
JVM version: 12.0.1
OS Information: Linux 4.15.0-50-generic
Usable space in artifacts repository: 19.81 GB
Database schema version: 1909001
Pipelines Count: 0
```

More details:
OS: Ubuntu 18.04 LTS. Fresh install, all available updated have been installed.
Apache installed as:
```
sudo apt install apache2
sudo a2enmod proxy 
sudo a2enmod proxy_http
```
Lets-encrypt certificates created and installed using `certbot` described here:
https://certbot.eff.org/lets-encrypt/ubuntubionic-apache

GO-Server, GoCD Server Version: 19.9.0. Installed from ubuntu packages as described here: https://docs.gocd.org/19.9.0/installation/install/server/linux.html

Reverse proxy setup as defined here https://docs.gocd.org/19.9.0/installation/configure-reverse-proxy.html

Installed github-oauth-authorization-plugin-3.0.2-57.jar into /var/lib/go-server/plugins/external

The plugin was configured using instructions here: https://github.com/gocd-contrib/github-oauth-authorization-plugin/blob/master/INSTALL.md

A dedicated github user was setup to own both the OAuth application in GitHub and to provide the Personal Access Token (rather than having the GitHub organisation own the OAuth application).

Other info:
- The only other authentication method enabled is a file-based password, with a single user.
- No agents or pilelines have yet been configured.

Many thanks in advance.

**Comment by andrewphilipsmith** [comment]
I have dug at little more and now have some more information about this:

1) I have now also attempted to setup the "google-oauth-authorization-plugin". (I do have a use case for using both auth plugins). I see almost identical issue - a long string which isn't the client ID, but looks like it could be an encrypted-representation of it being passed to the identity provider as the client_id parameter.

2) I tried enabling direct access to the GoCD server - ie not using Apache as a reverse proxy. (opened the relevant ports in the firewall, updated the siteUrls in the GoCD servers and updated the callback URLs from the identity providers). This did not result in any different behaviour or error messages.

3) I have found a possible workaround. *After* configuring both of the plugins, I opened the "Admin" -> "Config XML" page. For both plugings I replaced each of the encrypted values with the plaintext equivalent, changing the xml element from `encryptedValue` to `value`. eg:

Change from:
```
<authConfig id="github" pluginId="cd.go.authorization.github">
	<property>
		<key>ClientId</key>
		<encryptedValue>AES:long-random-string=</encryptedValue>
	</property>
</authConfig>
```
to:
```
<authConfig id="github" pluginId="cd.go.authorization.github">
	<property>
		<key>ClientId</key>
		<value>actual_client_id</value>
	</property>
</authConfig>
```
I repeated this for `ClientId`, `ClientSecret` and `PersonalAccessToken` for the github auth plugin and the `ClientId` and `ClientSecret` for the google auth plugin. I saved the the config xml and then rebooted the go-server.

After rebooting I am able to correctly login using both my github or google credentials. Additionally the strings are encrypted in config.xml.

I hope that it helpful info.

**Comment by ankitsri11** [comment]
@andrewphilipsmith I was able to replicate the behavior locally. I tried on a 19.9.0 GoCD instance with the plugin version 3.0.2-57. After the plugin setup, I got the GitHub's 404 error page for the user login.

I also tried your workaround and it worked as you have described.

My current suspicion is that it might be related to the way encryption values are stored in the cruise-config.xml from the plugin settings. It seems some extra characters are being added to the values.

I will check if someone from the team can look deeper. 

cc: @arvindsv

**Comment by akshaydewan** [comment]
I'm still looking into it, but it looks like some changes in [angular_plugin_new](https://github.com/gocd/gocd/commit/53440931d44ef84326dd6b1d3f3ad6c7efb7d6b6#diff-638b16bf8d2f93f6e3677f34e02122ddR91) might have caused this.

Reverting that line seems to fix the issue, but I still need to check why that change was made in the first place.

**Comment by denysvitali** [comment]
I can confirm that @andrewphilipsmith solution works, I didn't need to restart the server - the changes were applied after a couple of seconds 👍

**Comment by djmetzle** [comment]
I think this is fixed now upstream?

The change from https://github.com/gocd/gocd/pull/7108 shipped with [19.10.0](https://github.com/gocd/gocd/releases/tag/19.10.0)
That is also now included in the stable helm chart:
https://github.com/helm/charts/blob/139ee402d878ae6d84718393073997bf1b9594ca/stable/gocd/Chart.yaml#L5

---

:warning:  I had applied the workaround from https://github.com/gocd-contrib/github-oauth-authorization-plugin/issues/82#issuecomment-544467260 so i'm not directly confirming that the problem is actually fixed...
